### PR TITLE
[ConstantHoisting] Add a TTI hook to prevent hoisting.

### PR DESCRIPTION
**Pull request by paulwalker-arm**
Code generation can sometimes simplify expensive operations when
an operand is constant.  An example of this is divides on AArch64
where they can be rewritten using a cheaper sequence of multiplies
and subtracts.  Doing this is often better than hoisting expensive
constants which are likely to be hoisted by MachineLICM anyway.

### CODE REVIEW

**Comment by nikic** [comment]
Does this need to be target-specific? Not hoisting divisors to allow efficient expansion seems like something all targets would benefit from.

**Comment by davemgreen** [comment]
This should probably be in getIntImmCostInst, similar to the Arm version:
```
InstructionCost ARMTTIImpl::getIntImmCostInst(unsigned Opcode, unsigned Idx,
                                              const APInt &Imm, Type *Ty,
                                              TTI::TargetCostKind CostKind,
                                              Instruction *Inst) {
  // Division by a constant can be turned into multiplication, but only if we
  // know it's constant. So it's not so much that the immediate is cheap (it's
  // not), but that the alternative is worse.
  // FIXME: this is probably unneeded with GlobalISel.
  if ((Opcode == Instruction::SDiv || Opcode == Instruction::UDiv ||
       Opcode == Instruction::SRem || Opcode == Instruction::URem) &&
      Idx == 1)
    return 0;
```

**Comment by paulwalker-arm** [comment]
> Does this need to be target-specific? Not hoisting divisors to allow efficient expansion seems like something all targets would benefit from.

I guess not given the decision is derived from an existing TLI hook.  I've moved the implementation into BaseTTI.

> This should probably be in getIntImmCostInst, similar to the Arm version:

I originally considered this but it felt wrong because the code generation doesn't necessary result in cheap or free constants.  For the cases I want to handle the resulting constants are just as expensive with the gain being the removal of the divides.

**Comment by paulwalker-arm** [comment]
review ping

**Comment by davemgreen** [comment]
I've always considered the cost returned from getIntImmCostInst as the `cost(inst + imm)` - `cost(inst)`. So if we are producing a cheaper instruction, the cost of the imm can be cheap. That already happens in a few places, like saturating patterns under Arm and a few of the other patterns in other backends. Comparing a div vs multiple mul/adds is going to be a bit complex, but the immediate is also going to be completely different at any rate.

I'm not wanting to block this. If others disagree then feel free to go ahead, but this seems like the new interface might not really be needed.

**Comment by sjoerdmeijer** [comment]
I understood that hoisting and LICM were considered canonicalisation passes/transforms in LLVM. I have seen many attempts to restrict hoisting for one reason or another but I have seen them always being rejected because of the canonicalisation argument. Not sure this case is different, but perhaps @nikic who replied here can comment on that.

**Comment by nikic** [comment]
@sjoerdmeijer ConstaintHoisting is a backend pass, so it's fine to restrict it based on target-specific heuristics (in fact, the entire pass is already TTI-driven).

**Comment by paulwalker-arm** [comment]
I’ve looked to see if I can change the meaning of `getIntImmCostInst` to better reflect the intent of constant hoisting but I see it’s also used by ScalarEvolution when it calculates the cost of an expression tree.  I don’t see what else I can do differently here because at the end of the day the cost of generating the immediate for `sdiv i64 %a, 4294967087` is not free.

**Comment by paulwalker-arm** [comment]
Based on feedback from @david-arm I have renamed the TTI hook to `preferToKeepConstantsAttached` so that it doesn't give the false impression the instruction has anything to hoist.

**Review by david-arm — APPROVED** [review]
LGTM! After looking at the code and trying to understand the comments, I think overall this is the best approach. It looks like changing the meaning of getIntImmInstCost() will likely break other existing code.

File: llvm/include/llvm/Analysis/TargetTransformInfoImpl.h [hunk 1]
```diff
@@ -427,6 +427,11 @@ class TargetTransformInfoImplBase {
     return TTI::TCC_Free;
   }
 
+  bool preferToKeepConstantsAttached(const Instruction &Inst,
+                                     const Function &Fn) const {
+    return true;
```
Comment (nikic) [inline llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:432]: Shouldn't this default implementation return false rather than true?

Comment (paulwalker-arm) [reply llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:432]: Darn it. Thanks, I'll push an update.

**Review by nikic — APPROVED** [review]

**Pull request merged**